DOC improve documentation end-to-end pipeline (#999)
skrub-data/skrub: doc/end_to_end_pipeline.rst

.. currentmodule:: skrub

.. _tabular_learner:

Create baseline predictive models on heterogeneous dataset
----------------------------------------------------------

> Crafting a machine-learning pipeline is rather a daunting task. Choosing the ending

Reviewer: ```suggestion
Crafting a machine-learning pipeline is a rather daunting task. Choosing the ending
```

learner of such pipeline is usually the easiest part. However, it imposes constraints
regarding the preprocessing steps that are are required ahead of the learner.
Programmatically defining these steps is the part that requires the most expertise and
that is cumbersome to write.

The function :func:`tabular_learner` provides a factory function that given a
scikit-learn estimator, returns a pipeline that combines this estimator with the
appropriate preprocessing steps. Those steps corresponds to a :class:`TableVectorizer`
that is in charge of dealing with heterogeneous data and depending on the capabilities

> of the final estimator, a missing value imputer.

Reviewer: and/or a standard scaler?

Author: Right, it was added since I started this PR :)

> The function :func:`tabular_learner` provides a factory function that given a
> scikit-learn estimator, returns a pipeline that combines this estimator with the
> appropriate preprocessing steps. Those steps corresponds to a :class:`TableVectorizer`
> that is in charge of dealing with heterogeneous data and depending on the capabilities
> of the final estimator, a missing value imputer.

Reviewer: Nitpick: punctuation and slight rephrasing
```suggestion
The function :func:`tabular_learner` provides a factory function that, given a
scikit-learn estimator, returns a pipeline combining this estimator with the
appropriate preprocessing steps. These steps correspond to a :class:`TableVectorizer`
that handles heterogeneous data and, depending on the capabilities
of the final estimator, a missing value imputer.
```

either the native support of the model or the
:obj:`~sklearn.preprocessing.OrdinalEncoder`.

With linear models or unknown models, the default values of the different parameters are
used. Therefore, the :obj:`GapEncoder` is used for high-cardinality categorical features
and the :obj:`~sklearn.preprocessing.OneHotEncoder` for low-cardinality ones. If the
final estimator does not support missing values, a :obj:`~sklearn.impute.SimpleImputer`
is added before the final estimator. Finally, a
:obj:`~sklearn.preprocessing.StandardScaler` is added to the pipeline. Those choices may
not be optimal in all cases but they are methodologically safe.

.. _table_vectorizer:

Turning a dataframe into a numerical feature matrix
---------------------------------------------------

> A dataframe can contain columns of all kind of types. We usually refer such data to
> "heterogeneous" data. A good numerical representation of these columns help analytics
> and statistical learning.

Reviewer: ```suggestion
A dataframe can contain columns of all kinds of types. We usually refer to such data as
"heterogeneous" data. A good numerical representation of these columns helps with analytics
and statistical learning.
```

The :class:`TableVectorizer` gives a turn-key solution by applying different
data-specific encoders to the different columns. It makes reasonable heuristic choices
that are not necessarily optimal since it is not aware of the learner used for the

> machine learning task). However, it already provides a typically very good baseline.

Reviewer: ```suggestion
machine learning task. However, it already provides a typically very good baseline.
```

The :class:`TableVectorizer` handles the following type of data:

- numerical data represented with the data types `bool`, `int`, and `float`;
- categorical data represented with the data types `str` or categorical (e.g.
  :obj:`pandas.CategoricalDtype` or :obj:`polars.datatypes.Categorical`);
- date and time data represented by `datetime` data type (e.g. :obj:`numpy.datetime64`,
  :obj:`pandas.DatetimeTZDtype`, :obj:`polars.datatypes.Datetime`).

> The categorical data are subdivided into group: features containing a large number of
> categories (high-cardinality features) versus features containing a low number of
> categories (low-cardinality features). A feature is declared to be high-cardinality if
> the number of unique values is larger than a given threshold controlled by the
> parameter ``cardinality_threshold``.

Reviewer: Additional nitpicks
```suggestion
Categorical data are subdivided into two groups: features containing a large number of
categories (high-cardinality features) and features containing a small number of
categories (low-cardinality features). A feature is considered high-cardinality if
the number of unique values is greater than a given threshold, which is controlled by the
parameter cardinality_threshold.
```

either the native support of the model or the
:obj:`~sklearn.preprocessing.OrdinalEncoder`.

With linear models or unknown models, the default values of the different parameters are
used. Therefore, the :obj:`GapEncoder` is used for high-cardinality categorical features
and the :obj:`~sklearn.preprocessing.OneHotEncoder` for low-cardinality ones. If the
final estimator does not support missing values, a :obj:`~sklearn.impute.SimpleImputer`
is added before the final estimator. Finally, a
:obj:`~sklearn.preprocessing.StandardScaler` is added to the pipeline. Those choices may
not be optimal in all cases but they are methodologically safe.

.. _table_vectorizer:

> Turning a dataframe into a numerical feature matrix

Reviewer 2: ```suggestion
Turning a dataframe into a numeric feature matrix
```
